### PR TITLE
Ax 39 endpoint place pixel

### PR DESCRIPTION
**Pull request by NouCake**
Bitte mal drüberschauen :)

### CODE REVIEW

File: src/main/java/de/abstractolotl/azplace/AzplaceApplication.java [hunk 1]
```diff
@@ -2,9 +2,16 @@
...
+import org.springframework.context.annotation.Bean;
+import redis.clients.jedis.Jedis;
 
 @SpringBootApplication
 public class AzplaceApplication {
+    @Bean
+    public Jedis jedis() {
+        Jedis jedis = new Jedis("192.168.1.242", 6379);
+        return jedis;
+    }
```
Comment (Ancocodet) [inline src/main/java/de/abstractolotl/azplace/AzplaceApplication.java:14]: What is that for, i thougt we will work without a redis?

File: pom.xml [hunk 2]
```diff
@@ -17,39 +17,46 @@
...
+
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>redis.clients</groupId>
+            <artifactId>jedis</artifactId>
+        </dependency>
```
Comment (Ancocodet) [inline pom.xml:34]: I thougt we will work without redis?

File: src/main/java/de/abstractolotl/azplace/AzplaceApplication.java (hunk 1 above)
Comment (Ancocodet) [inline src/main/java/de/abstractolotl/azplace/AzplaceApplication.java:14]: I thougt we will work without redis

**Review by Ancocodet — CHANGES_REQUESTED** [review]

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    @Autowired private Jedis jedis;
+
+    @Autowired private CanvasRepo canvasRepo;
+    @Autowired private PixelOwnerRepo pixelOwnerRepo;
+
+    @PostMapping("/place")
+    private void place(@RequestBody PlaceRequest request) {
+        checkPixelCords(request.getX(), request.getY());
+        User user = getUserFromSession();
+        if(user == null) throw new RuntimeException("No user in Session"); //TODO
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java]: Why throwing a RuntimeException? Instead throw `new ResponseStatusException(HttpStatus.BAD_REQUEST, "No user in Session")`. This return the HttpStatusCode, i guess its easier to handle for the frondend.

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            throw new PixelOutOfBoundsException(request.getX(), request.getY(), canvas.getWidth(), canvas.getHeight());
+
+        setNewPixelOwner(canvas, request.getX(), request.getY(), getUserFromSession());
+        setPixelInBlob(canvas, request.getX(), request.getY(), request.getColor());
+    }
+
+    @GetMapping("/board/{canvasId}")
+    public byte[] boardData(@PathVariable int canvasId) {
+        var canvasRsp = canvasRepo.findById(canvasId);
+        if(canvasRsp.isEmpty()) throw new CanvasNotFoundExeption(canvasId);
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java:56]: Same here. At all endpoints i would recommend to throw HttpStatusCodes

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import redis.clients.jedis.Jedis;
+
+@RestController
+public class BoardController {
+
+    @Autowired private Jedis jedis;
+
+    @Autowired private CanvasRepo canvasRepo;
+    @Autowired private PixelOwnerRepo pixelOwnerRepo;
+
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java:23]: Swagger documentation missing

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        if(canvasResp.isEmpty()) throw  new CanvasNotFoundExeption(request.getCanvasId());
+
+        Canvas canvas = canvasResp.get();
+        if(canvas.getWidth() <= request.getX() || canvas.getHeight() <= request.getY())
+            throw new PixelOutOfBoundsException(request.getX(), request.getY(), canvas.getWidth(), canvas.getHeight());
+
+        setNewPixelOwner(canvas, request.getX(), request.getY(), getUserFromSession());
+        setPixelInBlob(canvas, request.getX(), request.getY(), request.getColor());
+    }
+
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java:46]: Swagger documentation missing

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import de.abstractolotl.azplace.model.requests.PlaceRequest;
+import de.abstractolotl.azplace.repositorys.CanvasRepo;
+import de.abstractolotl.azplace.repositorys.PixelOwnerRepo;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import de.abstractolotl.azplace.AzPlaceExceptions.*;
+import redis.clients.jedis.Jedis;
+
+@RestController
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java:16]: I would start the endpoint for all board requests with the prefix "board" so a request for place look like "localhost:8080/board/place".
Use: `@RestController("/board")
`

File: src/main/java/de/abstractolotl/azplace/AzplaceApplication.java [hunk 8]
```diff
@@ -2,9 +2,16 @@
...
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import redis.clients.jedis.Jedis;
 
 @SpringBootApplication
 public class AzplaceApplication {
+    @Bean
+    public Jedis jedis() {
+        Jedis jedis = new Jedis("192.168.1.242", 6379);
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/AzplaceApplication.java]: Avoid HardCoded IPs and Ports

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Autowired private Jedis jedis;
+
+    @Autowired private CanvasRepo canvasRepo;
+    @Autowired private PixelOwnerRepo pixelOwnerRepo;
+
+    @PostMapping("/place")
+    private void place(@RequestBody PlaceRequest request) {
+        checkPixelCords(request.getX(), request.getY());
+        User user = getUserFromSession();
```
Comment (jonas-be) [inline src/main/java/de/abstractolotl/azplace/controller/BoardController.java]: Discuss if we want to declare all immutable variables, as final.

**Review by jonas-be — CHANGES_REQUESTED** [review]
Rework needed

**Review by Ancocodet — APPROVED** [review]

File: src/main/java/de/abstractolotl/azplace/AzPlaceExceptions.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    public static class CanvasNotFoundExeption extends ResponseStatusException {
+        public CanvasNotFoundExeption(int canvasId) {
+            super(HttpStatus.I_AM_A_TEAPOT, "Could not find Canvas with ID: " + canvasId);
+        }
+    }
+
+    public static class PixelOutOfBoundsException extends ResponseStatusException {
+        public PixelOutOfBoundsException(int x, int y, int w, int h) {
+            super(HttpStatus.I_AM_A_TEAPOT, "Pixel coordinates are out of canvas bounds:\nCanvas: (" + w + ", " + h + ")\nRequested Pixel: (" + x + ", " + y + ")");
```
Comment (Ancocodet) [inline src/main/java/de/abstractolotl/azplace/AzPlaceExceptions.java]: Use the correct status code

File: src/main/java/de/abstractolotl/azplace/AzPlaceExceptions.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        }
+    }
+
+    public static class IllegalPixelCoordsException extends ResponseStatusException {
+        public IllegalPixelCoordsException() {
+            super(HttpStatus.I_AM_A_TEAPOT, "Pixel coordinates cannot be negative!");
```
Comment (Ancocodet) [inline src/main/java/de/abstractolotl/azplace/AzPlaceExceptions.java]: Use the correct status code

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -1,4 +1,9 @@
 springdoc.api-docs.path=/api-docs
 springdoc.swagger-ui.path=/swagger-ui
+server.port=13000
 
-
+spring.jpa.hibernate.ddl-auto=update
+spring.datasource.url=jdbc:mysql://noucake.ddns.net:3306/aztube
+spring.datasource.username=root
+spring.datasource.password=rootpw
```
Comment (Ancocodet) [inline src/main/resources/application.properties:8]: Passwords shouldnt be included inside the repository

File: src/main/resources/application.properties [hunk 13]
```diff
@@ -1,4 +1,9 @@
 springdoc.api-docs.path=/api-docs
 springdoc.swagger-ui.path=/swagger-ui
+server.port=13000
 
-
+spring.jpa.hibernate.ddl-auto=update
+spring.datasource.url=jdbc:mysql://noucake.ddns.net:3306/aztube
```
Comment (Ancocodet) [inline src/main/resources/application.properties:6]: db hostnames shouldnt be inside the repo

**Review by Ancocodet — CHANGES_REQUESTED** [review]

File: src/main/java/de/abstractolotl/azplace/AzplaceApplication.java (hunk 8 above)
Comment (Ancocodet) [reply src/main/java/de/abstractolotl/azplace/AzplaceApplication.java]: Use @Value instead

File: src/main/java/de/abstractolotl/azplace/controller/BoardController.java (hunk 5 above)
Comment (NouCake) [reply src/main/java/de/abstractolotl/azplace/controller/BoardController.java:21]: Für die Repos gibts swagger documentation?

File: src/main/java/de/abstractolotl/azplace/AzplaceApplication.java (hunk 8 above)
Comment (NouCake) [reply src/main/java/de/abstractolotl/azplace/AzplaceApplication.java]: https://aztube.atlassian.net/browse/AX-106

File: src/main/resources/application.properties (hunk 12 above)
Comment (NouCake) [reply src/main/resources/application.properties:8]: https://aztube.atlassian.net/browse/AX-106

**Pull request closed without merging**